File: apps/app/src/hooks/market-vision/rsi-divergences.ts
'use client'

import { findPairedDivergences, pivotHighAt, pivotLowAt } from './divergence-engine'
import type { OHLCVDataPoint, SeriesDataPoint } from './market-vision-config'
import {
  DEFAULT_REVERSE_RSI_TARGETS,
  type ReverseRsiLevel,
  ema,
  reverseRsiLevels,
  reverseRsiPrice,
  rsi as rsiCalc,
  sma,
  wilderRsiState,
} from './technical-indicators'

export type { ReverseRsiLevel }

export type RsiDivergenceType = 'bullish' | 'bearish' | 'h_bullish' | 'h_bearish'

export interface RsiDivergence {
  type: RsiDivergenceType
  startIndex: number
  endIndex: number
  startTime: number
  endTime: number
  rsiStart: number
  rsiEnd: number
  priceStart: number
  priceEnd: number
}

export interface RsiDivergencesConfig {
  rsiLength: number
  leftBars: number
  rightBars: number
  pairMode: 'TV-like' | 'Same Bar'
  tolBars: number
  priceMode: 'High/Low' | 'Close'
  allowEqual: boolean
  priceEps: number
  rsiEps: number
  showRegular: boolean
  showHidden: boolean
  /** Signal line period (moving average of RSI). */
  signalPeriod: number
  /** Signal line smoothing type. */
  signalType: 'EMA' | 'SMA'
  /** Alert level above the critical bull zone. */
  alertHigh: number
  /** Alert level below the critical bear zone. */
  alertLow: number
  /** Targets for reverse-RSI price levels (default: Caretaker zones 80/62/50/38/20). */
  reverseTargets?: readonly number[]
}

export interface RsiPivotPoint {
  index: number
  time: number
  value: number
  kind: 'high' | 'low'
}

// The Caretaker's zone levels drawn on the RSI pane.
export const RSI_ZONE_LEVELS = {
  critBull: 80,
  contBull: 62,
  middle: 50,
  contBear: 38,
  critBear: 20,
} as const

export interface RsiDivergencesResult {
  rsiSeries: SeriesDataPoint[]
  /** Moving average of RSI, masked past the RSI warm-up region. */
  signalSeries: SeriesDataPoint[]
  signalCurrent: number | null
  /** Next-bar close at which RSI crosses its signal line (null when unavailable). */
  reverseSignalCross: number | null
  /** Every confirmed RSI pivot high/low (same fractal as divergence pairing). */
  pivots: RsiPivotPoint[]
  levels: {
    critBull: SeriesDataPoint[]
    contBull: SeriesDataPoint[]
    middle: SeriesDataPoint[]
    contBear: SeriesDataPoint[]
    critBear: SeriesDataPoint[]
    alertHigh: SeriesDataPoint[]
    alertLow: SeriesDataPoint[]
  }
  alerts: { high: number; low: number; highOn: boolean; lowOn: boolean }
  divergences: RsiDivergence[]
  /** Next-bar close needed for RSI to print each target (null when unreachable). */
  reverseLevels: ReverseRsiLevel[]
}

const DEFAULT_CONFIG: RsiDivergencesConfig = {
  rsiLength: 14,
  leftBars: 5,
  rightBars: 5,
  pairMode: 'TV-like',
  tolBars: 2,
  priceMode: 'High/Low',
  allowEqual: true,
  priceEps: 0,
  rsiEps: 0,
  showRegular: true,
  showHidden: true,
  signalPeriod: 12,
  signalType: 'EMA',
  alertHigh: 85,
  alertLow: 15,
  reverseTargets: DEFAULT_REVERSE_RSI_TARGETS,
}

function buildLevels(times: number[], alertHigh: number, alertLow: number): RsiDivergencesResult['levels'] {
  return {
    critBull: times.map((time) => ({ time, value: RSI_ZONE_LEVELS.critBull })),
    contBull: times.map((time) => ({ time, value: RSI_ZONE_LEVELS.contBull })),
    middle: times.map((time) => ({ time, value: RSI_ZONE_LEVELS.middle })),
    contBear: times.map((time) => ({ time, value: RSI_ZONE_LEVELS.contBear })),
    critBear: times.map((time) => ({ time, value: RSI_ZONE_LEVELS.critBear })),
    alertHigh: times.map((time) => ({ time, value: alertHigh })),
    alertLow: times.map((time) => ({ time, value: alertLow })),
  }
}

/**
 * Signal line over the RSI. The RSI warm-up region (index 0 prints 0, indices
 * 1..rsiLength-1 print 100) must be sliced off before smoothing — ema() seeds
 * from the first value and would otherwise be poisoned for dozens of bars.
 */
function buildSignal(
  rsiValues: number[],
  times: number[],
  rsiLength: number,
  signalPeriod: number,
  signalType: 'EMA' | 'SMA',
): SeriesDataPoint[] {
  if (rsiValues.length <= rsiLength) return []

  const validRsi = rsiValues.slice(rsiLength)
  const smoothed = signalType === 'SMA' ? sma(validRsi, signalPeriod) : ema(validRsi, signalPeriod)
  // sma() zero-fills its first period-1 outputs; ema() is valid from index 0.
  const offset = signalType === 'SMA' ? signalPeriod - 1 : 0

  const points: SeriesDataPoint[] = []
  for (let i = offset; i < smoothed.length; i++) {
    const value = smoothed[i]
    const time = times[rsiLength + i]
    if (value == null || !Number.isFinite(value) || time == null) continue
    points.push({ time, value })
  }
  return points
}

/**
 * RSI value at which the next bar's RSI equals the next bar's signal.
 * EMA: nextSig = k*nextRsi + (1-k)*sigPrev, so nextRsi = nextSig ⇒ nextRsi = sigPrev.
 * SMA: nextSig = (sum of last period-1 RSI + nextRsi) / period ⇒ nextRsi = sum / (period-1).
 */
function signalCrossTarget(
  rsiValues: number[],
  rsiLength: number,
  signalPeriod: number,
  signalType: 'EMA' | 'SMA',
  signalCurrent: number | null,
): number | null {
  if (signalType === 'EMA') return signalCurrent

  if (signalPeriod < 2) return null
  const validRsi = rsiValues.slice(rsiLength)
  if (validRsi.length < signalPeriod - 1) return null
  const window = validRsi.slice(-(signalPeriod - 1))
  return window.reduce((sum, value) => sum + value, 0) / window.length
}

function findRsiPivots(
  rsiValues: number[],
  times: number[],
  leftBars: number,
  rightBars: number,
): RsiPivotPoint[] {
  const pivots: RsiPivotPoint[] = []
  for (let i = 0; i < rsiValues.length; i++) {
    const high = pivotHighAt(rsiValues, leftBars, rightBars, i)
    if (high != null) {
      pivots.push({ index: i - rightBars, time: times[i - rightBars] ?? 0, value: high, kind: 'high' })
    }
    const low = pivotLowAt(rsiValues, leftBars, rightBars, i)
    if (low != null) {
      pivots.push({ index: i - rightBars, time: times[i - rightBars] ?? 0, value: low, kind: 'low' })
    }
  }
  return pivots
}

export function calculateRsiDivergences(
  data: OHLCVDataPoint[],
  config?: Partial<RsiDivergencesConfig>,
): RsiDivergencesResult {
  const finalConfig: RsiDivergencesConfig = { ...DEFAULT_CONFIG, ...(config ?? {}) }

  if (!data.length) {
    return {
      rsiSeries: [],
      signalSeries: [],
      signalCurrent: null,
      reverseSignalCross: null,
      pivots: [],
      levels: { critBull: [], contBull: [], middle: [], contBear: [], critBear: [], alertHigh: [], alertLow: [] },
      alerts: { high: finalConfig.alertHigh, low: finalConfig.alertLow, highOn: false, lowOn: false },
      divergences: [],
      reverseLevels: [],
    }
  }

  const times = data.map((d) => d.time)
  const closes = data.map((d) => d.close)
  const highs = data.map((d) => (finalConfig.priceMode === 'High/Low' ? d.high : d.close))
  const lows = data.map((d) => (finalConfig.priceMode === 'High/Low' ? d.low : d.close))

  const rsiValues = rsiCalc(closes, finalConfig.rsiLength)

  const rsiSeries: SeriesDataPoint[] = times.map((time, i) => ({
    time,
    value: rsiValues[i] ?? 0,
  }))

  const paired = findPairedDivergences(highs, lows, rsiValues, {
    leftBars: finalConfig.leftBars,
    rightBars: finalConfig.rightBars,
    pairMode: finalConfig.pairMode,
    tolBars: finalConfig.tolBars,
    allowEqual: finalConfig.allowEqual,
    priceEps: finalConfig.priceEps,
    oscEps: finalConfig.rsiEps,
    showRegular: finalConfig.showRegular,
    showHidden: finalConfig.showHidden,
  })

  const divergences: RsiDivergence[] = paired.map((d) => ({
    type: d.type,
    startIndex: d.startIndex,
    endIndex: d.endIndex,
    startTime: times[d.startIndex] ?? 0,
    endTime: times[d.endIndex] ?? 0,
    rsiStart: d.oscStart,
    rsiEnd: d.oscEnd,
    priceStart: d.priceStart,
    priceEnd: d.priceEnd,
  }))

  const signalSeries = buildSignal(rsiValues, times, finalConfig.rsiLength, finalConfig.signalPeriod, finalConfig.signalType)
  const signalCurrent = signalSeries[signalSeries.length - 1]?.value ?? null

  const crossTarget = signalCrossTarget(
    rsiValues,
    finalConfig.rsiLength,
    finalConfig.signalPeriod,
    finalConfig.signalType,
    signalCurrent,
  )
  const crossState = crossTarget == null ? null : wilderRsiState(closes, finalConfig.rsiLength)
  const reverseSignalCross =
    crossState && crossTarget != null ? reverseRsiPrice(crossState, finalConfig.rsiLength, crossTarget) : null

  // Warm-up RSI prints 100 — never let it trip the overbought alert on short inputs.
  const lastRsi = closes.length > finalConfig.rsiLength ? (rsiValues[rsiValues.length - 1] ?? null) : null
  const alerts = {
    high: finalConfig.alertHigh,
    low: finalConfig.alertLow,
    highOn: lastRsi != null && lastRsi >= finalConfig.alertHigh,
    lowOn: lastRsi != null && lastRsi <= finalConfig.alertLow,
  }

  return {
    rsiSeries,
    signalSeries,
    signalCurrent,
    reverseSignalCross,
    pivots: findRsiPivots(rsiValues, times, finalConfig.leftBars, finalConfig.rightBars),
    levels: buildLevels(times, finalConfig.alertHigh, finalConfig.alertLow),
    alerts,
    divergences,
    reverseLevels: reverseRsiLevels(closes, finalConfig.rsiLength, finalConfig.reverseTargets),
  }
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Indicators/RsiDivergences.swift
import Foundation

/// Port of `rsi-divergences.ts`.
public enum RsiDivergences {
  public struct Config: Sendable, Hashable {
    public var rsiLength = 14
    public var leftBars = 5, rightBars = 5
    public var pairMode: PairMode = .tvLike
    public var tolBars = 2
    public var priceMode: PriceMode = .highLow
    public var allowEqual = true
    public var priceEps = 0.0, rsiEps = 0.0
    public var showRegular = true, showHidden = true
    public var signalPeriod = 12
    public var signalType: SignalType = .ema
    public var alertHigh = 85.0, alertLow = 15.0
    public var reverseTargets: [Double] = TechnicalIndicators.defaultReverseRsiTargets
    public init() {}
    public static let `default` = Config()
  }
  public enum PriceMode: String, Sendable, Codable { case highLow = "High/Low", close = "Close" }
  public enum SignalType: String, Sendable, Codable { case ema = "EMA", sma = "SMA" }

  public struct Divergence: Sendable, Hashable {
    public var type: DivergenceType
    public var startIndex: Int, endIndex: Int
    public var startTime: Int, endTime: Int
    public var rsiStart: Double, rsiEnd: Double, priceStart: Double, priceEnd: Double
  }

  public struct Pivot: Sendable, Hashable { public var index: Int; public var time: Int; public var value: Double; public var isHigh: Bool }

  public static let zoneLevels = (critBull: 80.0, contBull: 62.0, middle: 50.0, contBear: 38.0, critBear: 20.0)

  public struct Result: Sendable, Hashable {
    public var rsiSeries: [TimePoint]
    public var signalSeries: [TimePoint]
    public var signalCurrent: Double?
    public var reverseSignalCross: Double?
    public var pivots: [Pivot]
    public var alertHigh: Double, alertLow: Double, alertHighOn: Bool, alertLowOn: Bool
    public var divergences: [Divergence]
    public var reverseLevels: [TechnicalIndicators.ReverseRsiLevel]
    public var latestRsi: Double? { rsiSeries.last?.value }
  }

  static func buildSignal(_ rsi: [Double], _ times: [Int], rsiLength: Int, period: Int, type: SignalType) -> [TimePoint] {
    guard rsi.count > rsiLength else { return [] }
    let valid = Array(rsi[rsiLength...])
    let smoothed = type == .sma ? TechnicalIndicators.sma(valid, period) : TechnicalIndicators.ema(valid, period)
    let offset = type == .sma ? period - 1 : 0
    var out: [TimePoint] = []
    var i = offset
    while i < smoothed.count {
      if smoothed[i].isFinite, rsiLength + i < times.count { out.append(TimePoint(epochSeconds: times[rsiLength + i], value: smoothed[i])) }
      i += 1
    }
    return out
  }

  static func signalCrossTarget(_ rsi: [Double], rsiLength: Int, period: Int, type: SignalType, current: Double?) -> Double? {
    if type == .ema { return current }
    guard period >= 2, rsi.count > rsiLength else { return nil }
    let valid = Array(rsi[rsiLength...])
    guard valid.count >= period - 1 else { return nil }
    let window = valid.suffix(period - 1)
    return window.reduce(0, +) / Double(window.count)
  }

  public static func calculate(_ data: [OHLCVBar], config c: Config = .default) -> Result {
    guard !data.isEmpty else {
      return Result(rsiSeries: [], signalSeries: [], signalCurrent: nil, reverseSignalCross: nil, pivots: [], alertHigh: c.alertHigh, alertLow: c.alertLow, alertHighOn: false, alertLowOn: false, divergences: [], reverseLevels: [])
    }
    let times = data.map(\.time), closes = data.map(\.close)
    let highs = data.map { c.priceMode == .highLow ? $0.high : $0.close }
    let lows = data.map { c.priceMode == .highLow ? $0.low : $0.close }
    let rsi = TechnicalIndicators.rsi(closes, c.rsiLength)
    let rsiSeries = zip(times, rsi).map { TimePoint(epochSeconds: $0, value: $1) }
    let paired = DivergenceEngine.findPairedDivergences(highs: highs, lows: lows, osc: rsi, config: .init(leftBars: c.leftBars, rightBars: c.rightBars, pairMode: c.pairMode, tolBars: c.tolBars, allowEqual: c.allowEqual, priceEps: c.priceEps, oscEps: c.rsiEps, showRegular: c.showRegular, showHidden: c.showHidden))
    let divergences = paired.map { Divergence(type: $0.type, startIndex: $0.startIndex, endIndex: $0.endIndex, startTime: times[$0.startIndex], endTime: times[$0.endIndex], rsiStart: $0.oscStart, rsiEnd: $0.oscEnd, priceStart: $0.priceStart, priceEnd: $0.priceEnd) }
    let signal = buildSignal(rsi, times, rsiLength: c.rsiLength, period: c.signalPeriod, type: c.signalType)
    let signalCurrent = signal.last?.value
    let crossTarget = signalCrossTarget(rsi, rsiLength: c.rsiLength, period: c.signalPeriod, type: c.signalType, current: signalCurrent)
    let crossState = crossTarget == nil ? nil : TechnicalIndicators.wilderRsiState(closes, period: c.rsiLength)
    let reverseCross = (crossState != nil && crossTarget != nil) ? TechnicalIndicators.reverseRsiPrice(crossState!, period: c.rsiLength, target: crossTarget!) : nil
    let lastRsi: Double? = closes.count > c.rsiLength ? rsi.last : nil
    var pivots: [Pivot] = []
    for i in 0..<rsi.count {
      if let h = DivergenceEngine.pivotHighAt(rsi, leftBars: c.leftBars, rightBars: c.rightBars, currentIndex: i) { pivots.append(Pivot(index: i - c.rightBars, time: times[i - c.rightBars], value: h, isHigh: true)) }
      if let l = DivergenceEngine.pivotLowAt(rsi, leftBars: c.leftBars, rightBars: c.rightBars, currentIndex: i) { pivots.append(Pivot(index: i - c.rightBars, time: times[i - c.rightBars], value: l, isHigh: false)) }
    }
    return Result(rsiSeries: rsiSeries, signalSeries: signal, signalCurrent: signalCurrent, reverseSignalCross: reverseCross, pivots: pivots,
                  alertHigh: c.alertHigh, alertLow: c.alertLow, alertHighOn: (lastRsi ?? -1) >= c.alertHigh, alertLowOn: lastRsi != nil && lastRsi! <= c.alertLow,
                  divergences: divergences, reverseLevels: TechnicalIndicators.reverseRsiLevels(closes, period: c.rsiLength, targets: c.reverseTargets))
  }
}
